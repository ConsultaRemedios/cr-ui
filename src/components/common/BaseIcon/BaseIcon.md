**BaseIcon component** is responsible for rendering cr-ui icons. It uses the
[svg sprite concept](https://fvsch.com/code/svg-icons/) where you define, right after the body tag,
the svg icons with their respective ids (without rendering them) and import them where you want to
(now rendering it).

So, for rendering an icon using the `BaseIcon` component, all you need to do is passing
the reference (id) of the icon you defined previously as the id propriety.
The code would be something like:

#### HTML
```html
<body>
  <svg style="position: absolute; width: 0; height: 0">
    <symbol viewBox="0 0 24 24" id="favorite.icon">
      <path d="M16.4730679,3 C18.034356..."></path>
    </symbol>
  </svg>
  ...
</body>
```

#### JS
```js static
<BaseIcon id="'favorite.icon'">
```

## Using webpack and svg-sprite-loader
Building svg sprites manually is not a good choice. We suggest to use
[svg-sprite-loader](https://github.com/kisenka/svg-sprite-loader) on the projects which helps to
build and get the references of the icons.

Using `webpack` and `svg-sprite-loader`, the same example above,
could be written with just two lines of code - desconsidering the svg file.

```js static
import favorite from 'cr-ui/src/icons/favorite.icon.svg';
<BaseIcon :id="favorite.id">

```

## List of availables icons

```js
<div>
  <div v-for="icon in allIcons" :style="{ display: 'inline-block', padding: '15px 0', width: '33%' }">
    <BaseIcon :id="icon.id"/>
    <span>src/icons/{{icon.id}}.svg</span>
  </div>
</div>
```
