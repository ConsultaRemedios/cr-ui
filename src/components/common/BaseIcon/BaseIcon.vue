<template>
  <svg
    viewBox="0 0 24 24"
    :class="$style.icon"
    v-bind="$attrs"
  >
    <use :xlink:href="`#${id}`" />
  </svg>
</template>

<script>
export default {
  name: 'BaseIcon',
  inheritAttrs: false,
  props: {
    id: String,
  },
};
</script>

<style module>
.icon {
  display: inline-block;
  width: 1em;
  height: 1em;
  vertical-align: middle;
  fill: currentColor;
  font-size: 24px;
}
</style>
