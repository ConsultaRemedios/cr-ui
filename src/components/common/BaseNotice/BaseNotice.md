### Basic Usage

The BaseNotice works fully dinamically, the `title`, `message` and the `icons`. For `title` and `message` you should send simple strings. For
Icon, it works the same as `BaseIcon` component.
The component also has a `timeout` prop, it works with miliseconds, default is 5000, if you need more, just send as prop as we use in examples:

```
const getIcon = (icons, name) => icons.find((icon) => icon.id === `${name}.icon`) ;
<BaseNotice
  title="Something wrong happened"
  message="Lorem ipsum dolor sit amet"
/>
```
#### Types
```
const getIcon = (icons, name) => icons.find((icon) => icon.id === `${name}.icon`) ;

<div>
  <BaseNotice
    title="Something wrong happened"
    message="Lorem ipsum dolor!!"
    type="danger"
    :icon="getIcon(icons, 'warning')"
  />

  <br />

  <BaseNotice
    title="Sorry, you got an error!"
    message="Lorem ipsum dolor sit amet"
    type="warning"
    :icon="getIcon(icons, 'info')"
  />

  <br />

  <BaseNotice
    title="Congrats all works!"
    type="success"
    :icon="getIcon(icons, 'thumb-up')"
  />

  <br />

  <BaseNotice
    title="Something wrong happened"
    message="Lorem ipsum dolor?"
    type="info"
    :icon="getIcon(icons, 'help')"
  />


</div>
```
#### Snackbar

Defines the Notice type as snackbar, it should be `true` or `false`, default is `false`.
You should also use a method to control the Fading, see the example:

```js
new Vue({
  data() {
    return {
      error: '',
    }
  },
  methods: {
    setError() {
      this.error = 'Houve um erro, vocÃª precisa usar uma outra senha';
    },

    clearError() {
      this.error = '';
    },
  },

  template: `
    <div>
      <GridCol :span="{ md: 6 }" :offset="{ md: 4 }">
        <BaseNotice
          v-if="error"
          :snackbar="true"
          :title="error"
          type="danger"
          :scroll-to-self="false"
          :auto-close="true"
          @hide="clearError"
        />
      </GridCol>

      <BaseButton @click="setError">Mostrar Snackbar</BaseButton>
    </div>
  `
})
```
