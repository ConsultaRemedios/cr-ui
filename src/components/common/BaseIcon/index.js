import BaseIcon from './BaseIcon.vue';

export default BaseIcon;
