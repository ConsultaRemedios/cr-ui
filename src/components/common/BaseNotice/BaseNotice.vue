<template>
<transition
  :enter-class="$style.enter"
  :enter-active-class="$style.enterActive"
  :enter-to-class="$style.enterTo"
  :leave-class="$style.leave"
  :leave-active-class="$style.leaveActive"
  :leave-to-class="$style.leaveTo"
  @after-leave="$emit('hide')"
>
  <div
    v-show="show"
    :class="[{[$style.snackbar]: this.snackbar} , $style.notice, $style[type]]"
  >
    <ContainerBlock>
      <div :class="$style.content">
        <div :class="$style.iconBlock">
          <BaseIcon :id="icon.id" />
        </div>

       <div :class="$style.textBlock">
         <span :class="$style.title">
          {{title}}
          </span>

          <p v-if="message" :class="$style.message">
            {{message}}
          </p>
       </div>
      </div>
    </ContainerBlock>
  </div>
</transition>
</template>

<script>
  import ContainerBlock from './../ContainerBlock';
  import infoIcon from './../../../icons/info.icon.svg';
  import BaseIcon from '../BaseIcon';

  export default {
    name: 'BaseNotice',
    components: {
      ContainerBlock,
      BaseIcon,
    },

    data() {
      return {
        show: false,
      };
    },

    props: {
      type: {
        type: String,
        default: 'info',
        validator(value) {
          return ['info', 'warning', 'success', 'danger'].includes(value);
        },
      },

      title: {
        type: String,
        required: true,
      },

      message: {
        type: String,
      },

      scrollToSelf: {
        type: Boolean,
        default: true,
      },

      icon: {
        type: Object,
        default() { return infoIcon; },
      },

      timeout: {
        type: Number,
        default: 5000,
      },

      snackbar: {
        type: Boolean,
        default: false,
      },

      autoClose: {
        type: Boolean,
        default: false,
      },
    },

    mounted() {
      if (this.scrollToSelf && !this.snackbar) {
        window.scrollTo(0, this.$el.offsetTop);
      }

      this.show = true;

      if (this.autoClose) {
        setTimeout(() => {
          this.show = false;
        }, this.timeout);
      }
    },
  };
</script>

<style module>
  @value media-md, media-lg from './../../../styles/variables.css';

  .content {
    display: inline-flex;
    align-items: center;
  }

  .iconBlock {
      padding-right: 10px;
    }

  .notice {
    color: #fff;
    padding: 15px 0;
  }

  .icon {
    display: none;
  }

  .title {
    font-weight: 800;
  }

  .message {
    margin: 0;
  }

  .info {
    background: #aaaaaa;
  }

  .warning {
    background: #fb9c12;
  }

  .success {
    background: #1BB738
  }

  .danger {
    background: #D30B3F;
  }

  .snackbar {
    min-width: 300px;
    border-radius: 5px;
    z-index: 99999;
    position: fixed;
    top: 10px;
    left: 10px;
    right: 10px;
    box-shadow: 0 3px 4px 0 rgba(0,0,0,0.34);
  }

  .snackbar.info {
    background: rgb(170, 170, 170, 0.94);
  }

  .snackbar.warning {
    background: rgb(251, 156, 18, 0.94);
  }

  .snackbar.success {
    background: rgb(27, 183, 56, 0.94);
  }

  .snackbar.danger {
    background: rgba(211, 11, 63, 0.94);
  }

  .snackbar.enterActive,
  .snackbar.leaveActive {
    transition: opacity 0.2s ease-out,
                transform 0.2s ease-out;
  }

  .snackbar.enter,
  .snackbar.leaveTo {
    opacity: 0;
    transform: translateY(-100%);
  }

  .snackbar.enterTo,
  .snackbar.leave {
    opacity: 1;
    transform: translateY(0);
  }

  @media media-md {
    .snackbar {
      top: 20px;
      left: 50%;
      right: auto;
      transform: translateX(-50%);
      min-width: 680px;
    }

    .snackbar.enter,
    .snackbar.leaveTo {
      transform: translateX(-50%) translateY(-100%);
    }

    .snackbar.enterTo,
    .snackbar.leave {
      transform: translateX(-50%) translateY(0);
    }
  }

  @media media-lg {
    .snackbar {
      top: 95px;
      padding: 15px 10px;
    }

    .content {
      position: relative;
      align-items: center;
    }

    .iconBlock {
      padding-right: 17px;
    }

    .icon {
      display: block;
      position: absolute;
      top: 50%;
      left: 0;
      transform: translateY(-50%);
    }
  }

</style>
