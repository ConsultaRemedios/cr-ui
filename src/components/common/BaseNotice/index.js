import BaseNotice from './BaseNotice.vue';

export default BaseNotice;
