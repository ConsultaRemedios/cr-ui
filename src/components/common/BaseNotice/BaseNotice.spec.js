import { shallowMount } from '@vue/test-utils';
import warning from './../../../icons/warning.icon.svg';
import BaseIcon from './../BaseIcon';
import BaseNotice from './BaseNotice';

describe('BaseNotice component', () => {
  const types = ['info', 'warning', 'success', 'danger'];

  const mountComponent = ({ icon, type}) => {
    return shallowMount(BaseNotice, {
      propsData: {
        title: 'warning title!',
        message: 'warning message',
        icon,
        type,
      }
    });
  }

  types.forEach((type) => {
    it(`matches the snapshot when type is equal to ${type}`, async () => {
      const wrapper = mountComponent({ icon: warning, type });

      await wrapper.vm.$nextTick();

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('When there is no message', () => {
    it('matches the snapshot', async () => {
      const wrapper = shallowMount(BaseNotice, {
        propsData: {
          title: 'title for foo',
          icon: warning,
          type: 'warning',
        }
      });

      await wrapper.vm.$nextTick();

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('When there is no icon', () => {
    it('renders info icon', () => {
      const wrapper = shallowMount(BaseNotice, {
        propsData: {
          title: 'title for foo',
          message: 'message for foo',
          type: 'warning',
        }
      });

      expect(wrapper.findComponent(BaseIcon).vm.id).toEqual('info.icon.svg');
    });
  });

  describe('When is a snackbar', async () => {
    const wrapper = shallowMount(BaseNotice, {
      propsData: {
        title: 'VocÃª precisa corrigir os erros a baixo',
        snackbar: true,
      }
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.element).toMatchSnapshot();
  })
});
