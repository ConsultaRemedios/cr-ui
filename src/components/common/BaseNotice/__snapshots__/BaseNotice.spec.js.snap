// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseNotice component When is a snackbar matches the snapshot 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="snackbar notice info"
    style=""
  >
    <containerblock-stub>
      <div
        class="content"
      >
        <div
          class="iconBlock"
        >
          <baseicon-stub
            id="info.icon.svg"
          />
        </div>
         
        <div>
          <span
            class="title"
          >
            
            VocÃª precisa corrigir os erros a baixo
          
          </span>
           
          <!---->
        </div>
      </div>
    </containerblock-stub>
  </div>
</transition-stub>
`;

exports[`BaseNotice component When there is no message matches the snapshot 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="notice warning"
    style=""
  >
    <containerblock-stub>
      <div
        class="content"
      >
        <div
          class="iconBlock"
        >
          <baseicon-stub
            id="warning.icon.svg"
          />
        </div>
         
        <div>
          <span
            class="title"
          >
            
            title for foo
          
          </span>
           
          <!---->
        </div>
      </div>
    </containerblock-stub>
  </div>
</transition-stub>
`;

exports[`BaseNotice component matches the snapshot when type is equal to danger 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="notice danger"
    style=""
  >
    <containerblock-stub>
      <div
        class="content"
      >
        <div
          class="iconBlock"
        >
          <baseicon-stub
            id="warning.icon.svg"
          />
        </div>
         
        <div>
          <span
            class="title"
          >
            
            warning title!
          
          </span>
           
          <p
            class="message"
          >
            
            warning message
          
          </p>
        </div>
      </div>
    </containerblock-stub>
  </div>
</transition-stub>
`;

exports[`BaseNotice component matches the snapshot when type is equal to info 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="notice info"
    style=""
  >
    <containerblock-stub>
      <div
        class="content"
      >
        <div
          class="iconBlock"
        >
          <baseicon-stub
            id="warning.icon.svg"
          />
        </div>
         
        <div>
          <span
            class="title"
          >
            
            warning title!
          
          </span>
           
          <p
            class="message"
          >
            
            warning message
          
          </p>
        </div>
      </div>
    </containerblock-stub>
  </div>
</transition-stub>
`;

exports[`BaseNotice component matches the snapshot when type is equal to success 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="notice success"
    style=""
  >
    <containerblock-stub>
      <div
        class="content"
      >
        <div
          class="iconBlock"
        >
          <baseicon-stub
            id="warning.icon.svg"
          />
        </div>
         
        <div>
          <span
            class="title"
          >
            
            warning title!
          
          </span>
           
          <p
            class="message"
          >
            
            warning message
          
          </p>
        </div>
      </div>
    </containerblock-stub>
  </div>
</transition-stub>
`;

exports[`BaseNotice component matches the snapshot when type is equal to warning 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="notice warning"
    style=""
  >
    <containerblock-stub>
      <div
        class="content"
      >
        <div
          class="iconBlock"
        >
          <baseicon-stub
            id="warning.icon.svg"
          />
        </div>
         
        <div>
          <span
            class="title"
          >
            
            warning title!
          
          </span>
           
          <p
            class="message"
          >
            
            warning message
          
          </p>
        </div>
      </div>
    </containerblock-stub>
  </div>
</transition-stub>
`;
